ffi/src/primitives.rs

```diff
@@ -135,13 +210,36 @@ mod wasm {
 
                 fn into_repr_c(self, store: &mut Self::RustStore) -> SliceMut<$src> {
                     *store = self;
-                    SliceMut::from_slice(store)
+                    SliceMut::from_slice(Some(store))
                 }
                 unsafe fn try_from_repr_c(source: SliceMut<$src>, _: &mut ()) -> Result<Self> {
                     source.into_rust().ok_or(FfiReturn::ArgIsNull).map(|slice| slice.to_vec())
                 }
             }
 
+            impl CWrapperOutput<Vec<NonWasmIntPrimitive>> for Vec<$src> {
+                type ReturnType = Self;
+            }
+            impl COutPtr<Vec<NonWasmIntPrimitive>> for Vec<$src> {
+                type OutPtr = OutBoxedSlice<$src>;
+            }
+            impl COutPtrWrite<Vec<NonWasmIntPrimitive>> for Vec<$src> {
+                unsafe fn write_out(self, out_ptr: *mut Self::OutPtr) {
+                    let mut store = Vec::default();
+                    CTypeConvert::<Vec<NonWasmIntPrimitive>, _>::into_repr_c(self, &mut store);
+                    out_ptr.write(OutBoxedSlice::from_vec(Some(store)));
+                }
+            }
+            impl COutPtrRead<Vec<NonWasmIntPrimitive>> for Vec<$src> {
+                unsafe fn try_read_out(out_ptr: Self::OutPtr) -> Result<Self> {
+                    let slice = SliceMut::from_raw_parts_mut(out_ptr.as_mut_ptr(), out_ptr.len());
+                    let res = CTypeConvert::<Vec<NonWasmIntPrimitive>, _>::try_from_repr_c(slice, &mut ());
+                    let layout = Layout::new::<Vec<$src>>();
+                    __dealloc(out_ptr.as_mut_ptr().cast(), layout.size(), layout.align());
```

Comment:
Correct me if i miss something, but seems like we are trying to deallocate not memory vector points to, but vector itself, is this correct behavour?

Reply:
Yes, this is intended behavior

Reply:
yes, `Vec<T>` has to be deallocated, not `T` because(at the current moment?) `T` will never carry ownership. Specifically in this case it's about integers like `Vec<u8>`

Reply:
very good observation, I believe I fixed it, have a 2nd look please

Reply:
Seems fine now.
One small concern calling `allocate` on ZST is `UB` and i'm not sure if it's fine to call `deallocate` if `slice.len() == 0`.

Reply:
looking at the implementation of `Vec`, you are correct again. Will fix